### PR TITLE
Infer the framework using the TargetPlatformMoniker instead of TPI and TPV, GetReferenceNearestTargetFrameworkTask now handles P2P frameworks with platforms, Add a LockFile.GetTarget utility for aliases

### LINKED ISSUE #9894: Update `GetReferenceNearestTargetFrameworkTask()` to support target frameworks with platforms (such as net5.0-windows)

### LINKED ISSUE #9895: Use TargetFrameworkMoniker & TargetPlatformMoniker to infer the frameworks instead of using individual TFI, TFV, TPI, TPV properties
NuGet should use TargetFrameworkMoniker & TargetPlatformMoniker to infer the frameworks instead of using the individual TFI, TFV, TPI, TPV properties

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9895
Fixes: https://github.com/NuGet/Home/issues/9894
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
This is a loaded PR with many small changes. 

Things to look for: 

* NuGetFramework.ParseComponents uses TargetFrameworkMoniker and TargetPlatformMoniker instead of the 5 props. 
* Helper utility to detect the target framework now prefers TargetPlatformMoniker if passed, over TargetPlatformIdentifier/TargetPlatformMinVersion. 
* GetReferenceNearestTargetFrameworkTask can now handle P2Ps of TFMs with Platforms. 
* Added a `LockFile.GetTarget` utility method that can use an alias instead of a NuGetFramework. 

Fixed the tests + added a hack covered in https://github.com/NuGet/Home/issues/9913. 

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:

### CODE REVIEW

**Comment by nkolev92** [comment]
cc @dsplaisted @sfoslund

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs [hunk 1]
```diff
@@ -469,5 +469,161 @@ public void GetReferenceNearestTargetFrameworkTask_WithSingleTfmAndSingleAtfAndN
...
+
+
+        [Theory]
+        [InlineData(".NETFramework,Version=v4.7.2", "", "net46", ".NETFramework,Version=v4.6", "None", "net46")]
+        [InlineData(".NETCoreApp,Version=v5.0", "Windows,Version=7.0", "net5.0-windows", ".NETCoreApp,Version=v5.0", "Windows,Version=7.0", "net5.0-windows")]
+        [InlineData(".NETCoreApp,Version=v5.0", "Windows,Version=7.0", "net5.0", ".NETCoreApp,Version=v5.0", "None", "net5.0")]
+        [InlineData(".NETCoreApp,Version=v6.0", "Windows,Version=7.0", "net5.0;netcoreapp3.0", ".NETCoreApp,Version=v5.0;.NETCoreApp,Version=v3.0", "None;None", "net5.0")]
+        [InlineData(".NETCoreApp,Version=v6.0", "Android,Version=22.0", "net5.0;net5.0-android", ".NETCoreApp,Version=v5.0;.NETCoreApp,Version=v5.0", "None;android,Version=21.0", "net5.0-android")]
+        [InlineData(".NETCoreApp,Version=v5.0", null, "latestnet;latestnetstandard", ".NETFramework,Version=v4.7.2;.NETStandard,Version=v2.1", "None;None", "latestnetstandard")]
+        public void GetReferenceNearestTargetFrameworkTask_WithTargetFrameworkInformation_ReturnsCompatibleAlias(
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs:487]: @dsplaisted

Can you please look at this test and confirm this is the contract we agreed on? 

I am gonna more details in the PR body for how it's supposed to work.

Comment (dsplaisted) [reply test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs:487]: I think this matches what we discussed.

Comment (dsplaisted) [reply test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs:487]: I would probably add a test case for `net5.0` referencing a project multitargeted between `net5.0;net5.0-windows`.

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs [hunk 2]
```diff
@@ -469,5 +469,161 @@ public void GetReferenceNearestTargetFrameworkTask_WithSingleTfmAndSingleAtfAndN
...
+            testLogger.Errors.Should().Be(0);
+            testLogger.DebugMessages.Count.Should().Be(DEBUG_MESSAGE_COUNT_INPUT_OUTPUT);
+        }
+
+        [Theory]
+        [InlineData(".NETFramework,Version=v4.7.2", null, "net46;net472", ".NETFramework,Version=v4.6", "None")]
+        [InlineData(".NETFramework,Version=v4.7.2", null, "net46", ".NETFramework,Version=v4.6;.NETFramework,Version=v4.7.2", "None")]
+        [InlineData(".NETFramework,Version=v4.7.2", null, "net46;net472", ".NETFramework,Version=v4.6", "None;Windows,Version=77.0")]
+        public void GetReferenceNearestTargetFrameworkTask_WithInvalidParameters_Errors(
+    string currentProjectTFM, string currentProjectTPM, string refTargetFrameworks, string refTargetFrameworkMonikers, string refTargetPlatformMonikers)
```
Comment (dsplaisted) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs:564]: I'd add a case to this test for `net5.0` trying to reference `net5.0-windows` or `net5.0-windows;net5.0-androad`.

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs (hunk 1 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/GetReferenceNearestTargetFrameworkTaskTest.cs:487]: Good point, added.

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/MSBuildStaticGraphRestoreTests.cs [hunk 3]
```diff
@@ -869,6 +871,7 @@ public void GetTargetFrameworkInfos_WithCustomAliases_InfersCorrectTargetFramewo
                         { "TargetFrameworkMoniker", $"{FrameworkConstants.FrameworkIdentifiers.Net},Version=4.6.1" },
                         { "TargetPlatformIdentifier", "" },
                         { "TargetPlatformVersion", "" },
+                        { "TargetPlatformVersion", "" },
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/MSBuildStaticGraphRestoreTests.cs]: I think you meant Moniker?
```suggestion
                        { "TargetPlatformMoniker", "" },
```

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs [hunk 4]
```diff
@@ -3206,6 +3255,7 @@ public void MSBuildRestoreUtility_AddPackageDownloads_SinglePackageSingleVersion
                         { "TargetFrameworkVersion", $"v{targetFramework.Version.ToString(2)}" },
                         { "TargetFrameworkMoniker", $"{targetFramework.Framework},Version={targetFramework.Version.ToString(2)}" },
                         { "TargetPlatformIdentifier", "" },
+                    { "TargetPlatformMoniker", "" },
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs:3258]: formatting?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs [hunk 5]
```diff
@@ -3261,6 +3311,7 @@ public void MSBuildRestoreUtility_AddPackageDownloads_SinglePackageMultipleVersi
...
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs:3314]: formatting?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs [hunk 6]
```diff
@@ -3322,6 +3373,7 @@ public void MSBuildRestoreUtility_AddPackageDownloads_InvalidVersionRange(string
...
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs:3376]: formatting?

**Review by zivkan — COMMENTED** [review]
I think you need to rebase on the latest dev, because I expect changes needed to VS nomination utilities + tests as well.

File: test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/MSBuildStaticGraphRestoreTests.cs (hunk 3 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/MSBuildStaticGraphRestoreTests.cs]: yep...fixed that in a future commit. The change had stayed in my local changes.

File: src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs [hunk 7]
```diff
@@ -73,9 +73,9 @@ public static NuGetFramework Parse(string folderName, IFrameworkNameProvider map
         /// <summary>
         /// Creates a NuGetFramework from individual components
         /// </summary>
-        public static NuGetFramework ParseComponents(string targetFrameworkIdentifier, string targetFrameworkVersion, string targetFrameworkProfile, string targetPlatformIdentifier, string targetPlatformVersion)
+        public static NuGetFramework ParseComponents(string targetFrameworkMoniker, string targetPlatformMoniker)
```
Comment (zkat) [inline src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs:76]: What happened to MinVersion?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs [hunk 8]
```diff
@@ -9978,7 +9978,7 @@ public void RestoreNetCore_WithMultipleFrameworksWithPlatformAndAssetTargetFallb
                 var frameworks = new NuGetFramework[]
                 {
                     NuGetFramework.Parse("net5.0-windows10.0.10000.1"),
-                    NuGetFramework.Parse("net5.0-android21")
+                    NuGetFramework.Parse("net5.0-android21.0")
```
Comment (zkat) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: I thought the . wasn't required for platformversion?

File: src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs (hunk 7 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs:76]: It was never used in the framework parsing.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs (hunk 6 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs:3376]: fixed, all formatting issues.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs (hunk 8 above)
Comment (dsplaisted) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: It shouldn't be required.  https://github.com/NuGet/Home/issues/9842

Comment (nkolev92) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: It's not...it's a limitation of the testing infra I am using...the condition added there is the string passed here, and it needs to match.

Comment (nkolev92) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: fixed the infra..should be better now.

**Comment by nkolev92** [comment]
Fixed all tests + added a property to help tell MSBuild that the new GetNearest Task supports TPM. 

cc @dsplaisted

**Comment by nkolev92** [comment]
@zkat @zivkan 

Can you please review? 

https://github.com/NuGet/Home/issues/9914 would depend on this.

File: src/NuGet.Core/NuGet.Build.Tasks/GlobalSuppressions.cs [hunk 9]
```diff
@@ -84,3 +82,4 @@
...
 [assembly: SuppressMessage("Build", "CA1052:Type 'RestoreSettingsUtils' is a static holder type but is neither static nor NotInheritable", Justification = "<Pending>", Scope = "type", Target = "~T:NuGet.Build.Tasks.RestoreSettingsUtils")]
 [assembly: SuppressMessage("Build", "CA1063:Provide an overridable implementation of Dispose(bool) on 'RestoreTask' or mark the type as sealed. A call to Dispose(false) should only clean up native resources. A call to Dispose(true) should clean up both managed and native resources.", Justification = "<Pending>", Scope = "type", Target = "~T:NuGet.Build.Tasks.RestoreTask")]
+[assembly: SuppressMessage("Globalization", "CA1307:Specify StringComparison", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.Build.Tasks.GetReferenceNearestTargetFrameworkTask.TryParseFramework(System.String,System.String,System.String,NuGet.Build.MSBuildLogger,NuGet.Frameworks.NuGetFramework@)~System.Boolean")]
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks/GlobalSuppressions.cs:85]: provide a justification?

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets [hunk 10]
```diff
@@ -57,6 +57,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     <NuGetInteractive Condition=" '$(NuGetInteractive)' == '' ">$(MSBuildInteractive)</NuGetInteractive>
     <!-- Mark that this targets file supports package download. -->
     <PackageDownloadSupported>true</PackageDownloadSupported>
+    <!-- Mark that this targets file GetReferenceNearestTargetFrameworkTask task supports the TargetPlatformMoniker -->
+    <GetReferenceNearestTargetFrameworkTaskSupportsTargetPlatformParameter>true</GetReferenceNearestTargetFrameworkTaskSupportsTargetPlatformParameter>
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:61]: start with an underscore, to use the convention of "internal" names?  I'm guessing we don't want customers depending on this property long term.

**Review by zivkan — APPROVED** [review]
Just minor nitpicks

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets (hunk 10 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:61]: It's technically used cross products.

File: src/NuGet.Core/NuGet.Build.Tasks/GlobalSuppressions.cs (hunk 9 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/GlobalSuppressions.cs:85]: do you mind if I do it in a future PR? 

Don't want to run another CI if there's no additional feedback.

Comment (zivkan) [reply src/NuGet.Core/NuGet.Build.Tasks/GlobalSuppressions.cs:85]: sure, no problem.

**Pull request merged**